File: majorityElement/majority1.java
//solution1: O(n), O(n) using hashmap
public int majority(int[] array) {
    // Write your solution here.
    return 0;
}
//solution2: O(n), O(1) using k-1 counter

// map<candidate, counter>
// case1: counter==0,counter++, update candidate
// case2.1: if cur == candidate, counter++
// case2.2: if not, counter--

public int majority(int[] arr) {
  int candidate = arr[0];
  int counter = 1;
  for(int i = 1; i < arr.length; i++){
    if(counter == 0){
      counter++;
      candidate = arr[i];
    }else if(candidate == arr[i]){
      counter++;
    }else{
      counter--;
    }
  }
  return candidate;
}

File: hashtable/majorityElement/majority1.py
# the majority number is the number that occurs more than half of the size of the array.
# soloution1: hashmap   O(n) O(n)

# solution2: k-1 counter O(n)  O(1)
# // map<candidate, counter>
# // case1: counter==0,counter++, update candidate
# // case2.1: if cur == candidate, counter++
# // case2.2: if not, counter--
def majority1(arr):
    res = None
    counter = 1
    for each in arr:
        if each == res:
            counter += 1
        else:
            counter -= 1
            if counter == 0:
                res = each
                counter = 1

    return res


# Boyer–Moore majority vote algorithm
